components/sinks/cu_zenoh_ros_sink/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+mod liveliness;
+mod node;
+mod topic;
+
+use attachment::encode_attachment;
+use cdr::{CdrBe, Infinite};
+use cu29::clock::RobotClock;
+use cu29::prelude::*;
+use cu_ros_payloads::RosMsgAdapter;
+use error::cu_error_map;
+use liveliness::{node_liveliness, publisher_liveliness};
+use node::Node;
+use topic::Topic;
+use zenoh::bytes::Encoding;
+use zenoh::Config;
+
+use std::marker::PhantomData;
+
+// Only one node per session
+const NODE_ID: u32 = 0;
+
+// Only one publisher per session
+const PUBLISHER_ID: u32 = NODE_ID + 1;
+
+/// This is a sink task that sends ROS-compatible messages to a Zenoh topic.
+/// P is the payload type of the messages, which must be convertible to ROS format.
+/// Hence the payload type must implement the `RosMsgAdapter` trait.
+pub struct ZenohRosSink<P>
+where
+    P: CuMsgPayload + RosMsgAdapter<'static>,
+{
+    _marker: PhantomData<P>,
+    config: ZenohRosConfig,
+    ctx: Option<ZenohRosContext>,
+}
+
+pub struct ZenohRosConfig {
+    session: zenoh::Config,
+    domain_id: u32,
+    namespace: String,
+    node: String,
+    topic: String,
+}
+
+pub struct ZenohRosContext {
+    session: zenoh::Session,
+    publisher: zenoh::pubsub::Publisher<'static>,
+    // Token have to be kept alive (undeclared on drop)
+    #[allow(dead_code)]
+    node_token: zenoh::liveliness::LivelinessToken,
+    #[allow(dead_code)]
+    publisher_token: zenoh::liveliness::LivelinessToken,
+    sequence_number: u64,
+}
+
+impl<P> Freezable for ZenohRosSink<P> where P: CuMsgPayload + RosMsgAdapter<'static> {}
+
+impl<'cl, P> CuSinkTask<'cl> for ZenohRosSink<P>
+where
+    P: CuMsgPayload + RosMsgAdapter<'static> + 'cl,
+{
+    type Input = input_msg!('cl, P);
+
+    fn new(config: Option<&ComponentConfig>) -> CuResult<Self>
+    where
+        Self: Sized,
+    {
+        let config = config.ok_or(CuError::from("ZenohRosSink: Missing configuration"))?;
+
+        // Get json zenoh config
+        let session_config = config.get::<String>("zenoh_config_json").map_or(
+            // Or default zenoh config otherwise
+            CuResult::Ok(Config::default()),
+            |s| -> CuResult<zenoh::Config> {
+                Config::from_json5(&s)
+                    .map_err(cu_error_map("ZenohRosSink: Failed to create zenoh config"))
+            },
+        )?;
+
+        Ok(Self {
+            _marker: Default::default(),
+            config: ZenohRosConfig {
+                session: session_config,
+                domain_id: config.get::<u32>("domain_id").unwrap_or(0),
+                namespace: config.get::<String>("namespace").unwrap_or("node".into()),
+                node: config.get::<String>("node").unwrap_or("node".into()),
+                topic: config.get::<String>("topic").unwrap_or("copper".into()),
+            },
+            ctx: None,
+        })
+    }
+
+    fn start(&mut self, _clock: &RobotClock) -> CuResult<()> {
+        let session = zenoh::Wait::wait(zenoh::open(self.config.session.clone()))
```

Comment:
How heavy it’s to create a new session per task, instead of having somehow a global ?

Reply:
in a separated projet using copper I have instantiated within a lazy operator

Reply:
This sounds great. However I'm wondering how to define a single zenoh session config for all the tasks inside the RON file? :thinking: 
@gbin any thoughts about this?

Reply:
I think we really need a new kind of "service" task just holding a state / running a daemon thread. I propose to add that as an issue for that and discuss the arch and determinism on the issue. For now let's move forward with this PR with the known limitation then refactor it with the new facility in the framework.

Reply:
really good idea. Looking forward to it